Extend combined scheme to Schemes in Type

TheoWinterhalter opened the pull request:
<!-- Thank you for your contribution.
     Make sure you read the contributing guide and fill this template. -->


<!-- Keep what applies -->
**Kind:** feature.


Extends command `Combined Scheme` to handle induction schemes that were generated in sort different from `Prop` (though it only checks the sort of the first predicate).
For this I modified `library/coqlib.ml` to give access to `prod` and `pair` but I didn't know where to put them within the file.
I should probably clean it and squash it before merging.

When the sort isn't `Prop`, it replaces `conj` and `and` by `pair` and `prod` in the generated term/type, meaning the cases that were already working with `Prop` remain unchanged.


**Fix: #6588**


<!-- If this is a feature pull request / breaks compatibility: -->
<!-- (Otherwise, remove these lines.) -->
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in CHANGES.

mattam82 commented on vernac/indschemes.ml:
```diff
@@ -460,12 +463,39 @@ let build_combined_scheme env schemes =
+  (* Return true iff the target sort is Prop *)
+  let is_sort_prop ty =
+    let (ctx, arity) = decompose_prod ty in
+    (* arity is the predicate applied to arguments, so we get its indice
```
Typo: index

mattam82 commented on vernac/indschemes.ml:
```diff
@@ -460,12 +463,39 @@ let build_combined_scheme env schemes =
+      | App (t, args) ->
+        begin match Constr.kind t with
+        | Rel i -> i - 1
+        | _ -> assert false
```
You should rather use anomaly to report where and what is not supported.

TheoWinterhalter commented:
I pushed an update with fixes to the remarks. Hopefully I used anomalies in the correct way.

I also feel I should be guided in where to place the changes I made to `coqlib.ml/i`.

mattam82 commented on library/coqlib.mli:
```diff
@@ -206,3 +206,8 @@ val coq_sig_ref : GlobRef.t lazy_t
 val coq_iff_ref : GlobRef.t lazy_t
+
+
+(** TODO Move *)
```
I'd move them under the conjunction(s) part

TheoWinterhalter commented:
Squashed commits.

maximedenes commented:
@mattam82 can you self-assign?

SkySkimmer commented on vernac/indschemes.ml:
```diff
@@ -460,12 +463,39 @@ let build_combined_scheme env schemes =
 	| _ -> ctx, destInd last, 0
   in
+  (* Return true iff the target sort is Prop *)
+  let is_sort_prop ty =
```
I think this function can be simplified to `Retyping.get_sort_family_of ty == Sorts.InProp`.

TheoWinterhalter commented:
Ok, I just read the test-suite/README which basically said so, as this is not really a bug fix. Please guide me.

SkySkimmer commented:
>Ok, I just read the test-suite/README which basically said so

It's certainly a bit of a mess.

Since it closes an issue it can be considered a bug, or you can add it to the success/ tests, eg success/Inductive.v

JasonGross commented:
I think the thing to check is that it works / builds a well-typed scheme, in which case the test should go in success/ rather than output/, and the Prints can be dropped

maximedenes commented:
I would check convertibility against the expected scheme.

TheoWinterhalter commented:
Thank you, I hope the test I provided is better suited.

maximedenes commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -14,7 +14,7 @@ syntax follows the schema:
 .. cmd:: Scheme @ident := Induction for @ident Sort sort {* with @ident := Induction for @ident Sort sort}
 
-where each `ident'ᵢ` is a different inductive type identifier 
+where each `ident'ᵢ` is a different inductive type identifier
```
Indentation seems wrong here (it was aready before this PR, but while we are at it...)

mattam82 commented:
I think the CI failures are spurious, can anyone confirm? @SkySkimmer maybe?

SkySkimmer commented:
The compcert failure is because this PR needs rebasing.
The others were runner issues, and can't be retried because artifacts were purged (pipeline is >1week old). They'll be run after the rebase regardless.

mattam82 commented:
The failures look spurious, @TheoWinterhalter can you rebase again resolving the conflict in CHANGES? I'll merge afterwards.

TheoWinterhalter commented:
Done.

mattam82 commented:
@TheoWinterhalter there's a conflict in proof-schemes now, please rebase

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -44,9 +44,9 @@ induction for objects in type `identᵢ`.
-     Axiom B : Set.	       
-   
+     Axiom A : Set.
+     Axiom B : Set.
```
Bad indentation.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -145,13 +145,16 @@ schema :
-This command generates the leftmost `ident` to be the conjunction of the
-principles: it is built from the common premises of the principles and
-concluded by the conjunction of their conclusions.
+  where each identᵢ after the ``from`` is a different inductive principle that must
```
It would be best to reformulate this paragraph so that the description comes entirely after the command grammar.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -173,6 +176,23 @@ concluded by the conjunction of their conclusions.
 
+.. example::
+
+  We can also combine schemes at sort ``Type``:
```
Aligning the beginning of the paragraph with the word `example` would be the expected indentation here (similarly in the following with `coqtop`).

TheoWinterhalter commented:
@Zimmi48: I haven't squashed or rebased for the time being in order to let you review the new changes I made to the doc.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -12,13 +12,15 @@ The ``Scheme`` command is a high-level tool for generating automatically
 syntax follows the schema:
 
-.. cmd:: Scheme @ident := Induction for @ident Sort sort {* with @ident := Induction for @ident Sort sort}
+.. cmd:: Scheme @ident := Induction for @ident' Sort sort {* with @ident := Induction for @ident Sort sort}
```
`@ident` supports subscript `__1` `__2` ... (notice the two `_`).
You could write for instance `Scheme @ident__1 := Induction for @ident__2 Sort sort {* with @ident__i := Induction for @ident__j Sort sort}`

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -12,13 +12,15 @@ The ``Scheme`` command is a high-level tool for generating automatically
-induction for objects in type `identᵢ`.
+  This command is a high-level tool for generating automatically
+  (possibly mutual) induction principles for given types and sorts.
+  Each `ident'ᵢ` is a different inductive type identifier belonging to
```
Use ```:n:`@ident__j` ``` Single backticks do not work alone.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -12,13 +12,15 @@ The ``Scheme`` command is a high-level tool for generating automatically
+  (possibly mutual) induction principles for given types and sorts.
+  Each `ident'ᵢ` is a different inductive type identifier belonging to
+  the same package of mutual inductive definitions.
+  The command generates the `identᵢ`s to be mutually recursive
```
You can use ```:n:`@ident__i`\s` ``` to put an `s` without spacing after the quoted text.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -139,19 +141,20 @@ Automatic declaration of schemes
-schema :
-
-.. cmd:: Combined Scheme @ident from {+, ident}
+.. cmd:: Combined Scheme @ident from {+, ident'}
```
You forgot an `@`. And same suggestion as above.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -139,19 +141,20 @@ Automatic declaration of schemes
-principles: it is built from the common premises of the principles and
-concluded by the conjunction of their conclusions.
+  This command is a tool for combining induction principles generated
+  by the ``Scheme command``.
```
```by the :cmd:`Scheme` command```.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -316,15 +336,15 @@ identᵢ’.
 .. cmd:: Derive Inversion @ident with forall (x : T), I t Sort sort
 
-This command generates an inversion principle for the `inversion … using` 
+This command generates an inversion principle for the `inversion … using`
```
This description is not correctly indented. It would be best to reformulate the paragraph so that the sentence ```The syntax of ``Derive`` ``Inversion`` follows the schema:``` disappears.

Zimmi48 commented:
It's not. We have trouble with bedrock these days.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -12,13 +12,15 @@ The ``Scheme`` command is a high-level tool for generating automatically
+  (possibly mutual) induction principles for given types and sorts.
+  Each :n:`ident__j` is a different inductive type identifier belonging to
+  the same package of mutual inductive definitions.
+  The command generates the :n:`ident__i`\s` to be mutually recursive
```
There's a spurious backtick *after* the `s`

TheoWinterhalter replied:
It actually came from you haha. ;)

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -136,19 +138,20 @@ Automatic declaration of schemes
-This command generates the leftmost `ident` to be the conjunction of the
-principles: it is built from the common premises of the principles and
-concluded by the conjunction of their conclusions.
+  This command is a tool for combining induction principles generated
```
Ideally, I prefer indenting with three spaces even if (I think) this is technically OK with two spaces too.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -12,13 +12,15 @@ The ``Scheme`` command is a high-level tool for generating automatically
-induction for objects in type `identᵢ`.
+  This command is a high-level tool for generating automatically
+  (possibly mutual) induction principles for given types and sorts.
+  Each :n:`ident__j` is a different inductive type identifier belonging to
```
When you use `:n:` the `ident` must be preceded with `@`.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -313,20 +333,18 @@ identᵢ’.
+  This command generates an inversion principle for the `inversion … using`
+  tactic. Let `I` be an inductive predicate and `x` the variables occurring
+  in t. This command generates and stocks the inversion lemma for the
+  sort `sort` corresponding to the instance `∀ (x:T), I t` with the name
```
Use `:g:` for quoting Coq terms. Otherwise, "normal" verbatim takes two backticks before and after.

Zimmi48 commented on doc/sphinx/user-extensions/proof-schemes.rst:
```diff
@@ -313,20 +333,18 @@ identᵢ’.
-sort `sort` corresponding to the instance `∀ (x:T), I t` with the name
-`ident` in the global environment. When applied, it is equivalent to
-having inverted the instance with the tactic `inversion`.
+  This command generates an inversion principle for the `inversion … using`
```
It would be best to write this ```:tacn:`inversion … using` ``` to have a direct link to the doc. This is not possible yet, because this tactic has no index entry. You can add one with the `:name:` option. (See the example of `induction ... using ...` in the same Tactics chapter.)

Zimmi48 commented:
I realize that some of my comments are on parts of the doc that you only re-indented. Feel free to ignore them if that's too much (although following them would make this part of the doc better, which would be cool).

TheoWinterhalter commented:
I tried to account for the changes you requires, squashed and rebased.
I can't seem to be able to build the doc at the moment so I'm doing it in blind mode, sorry if I have pushed mistakes again.

ejgallego commented:
> I can't seem to be able to build the doc at the moment so I'm doing it in blind mode, sorry if I have pushed mistakes again.

I'm afraid you need to use very concrete Python package versions [as outlined in Coq's Dockerfile]

I am regretting big time not opposing raising documentation dependencies, but in the future I will.

herbelin commented:
> I can't seem to be able to build the doc at the moment so I'm doing it in blind mode, sorry if I have pushed mistakes again.

Me neither actually. On MacOS X, brew does not seem to have the sphinx versions required by Coq yet.

ejgallego commented:
> Me neither actually. On MacOS X, brew does not seem to have the sphinx versions required by Coq yet.

@herbelin I'm afraid that you may need to `pip/pip3` then.

herbelin commented:
When I pip3, I get:
```
Requirement already satisfied: sphinx in /usr/local/lib/python3.6/site-packages (1.7.2)
...
```
But the `README.rst` says that sphinx >= 1.7.8 is needed.

ejgallego commented:
> But the `README.md` says that sphinx >= 1.7.8 is needed.

See the line in the `Dockerfile`:ç
```
pip3 install sphinx==1.7.8 sphinx_rtd_theme==0.2.5b2 \
                 antlr4-python3-runtime==4.7.1 sphinxcontrib-bibtex==0.4.0
```

herbelin commented:
Many thanks. It worked!

Zimmi48 commented:
I'm surprised that brew doesn't have these versions already. Anyways, using older versions works as well. It's just that if you have Sphinx < 1.7.0, you will get a few spurious warnings so you must make them non-fatal.

Zimmi48 commented:
@TheoWinterhalter The ```:tacn:`inversion … using` ``` thing is not working because you would need to also add the reference in the tactics chapter. Anyways, just undo this change if that's too much.

TheoWinterhalter commented:
I added the `:name:` option to `tactics.rst`, I don't know if that is sufficient.

Zimmi48 commented:
@TheoWinterhalter Not yet, make sure you use the exact same name. Currently, your `tacn` uses "inversion … using" and your `:name:` uses "inversion ... using ...". There are two difference single-character vs multi-character ... and the presence or the absence of such ... after using.

TheoWinterhalter commented:
Sorry. I assume following the version used by induction is the right way.
I guess I should learn how to use phinx properly.
Anyway hope this does the trick this time.

Zimmi48 commented:
This worked this time. Here is the rendered output: https://coq.gitlab.io/-/coq/-/jobs/101702366/artifacts/_install_ci/share/doc/coq/sphinx/html/user-extensions/proof-schemes.html

mattam82 commented:
Thanks!

gasche commented:
Thanks Theo! I'm slightly sad because I was planning to make a small contribution to Coq by implementing this feature, but I'm also happy to get it for free :-)

gasche commented:
P.S.: Hopefully one day this version of "Combined Scheme" will become the default family of induction/recursion principles generated for mutually-inductive definitions?

TheoWinterhalter commented:
Well, there is still hope, my improvement is still not complete. There are cases were it fails because of `decompose_prod` errors (but this is not related to adding `Prop` as en elim sort).

It would be better to generate the principles using parametricity and thus have something strong enough to deal with inductive inductive types.

herbelin commented:
> There are cases were it fails because of `decompose_prod` errors.

Can you say more? Maybe this is easy to fix.

mattam82 commented:
The invariant of combined scheme is that all principles have the same methods. In some cases the mutual induction principle generation is a bit too smart and generates a stronger principle because it can see that some things are independent, that can result in this issue when we try to combine them.
I have a PR somewhere that optionally disables this behavior of mutual scheme generation.
